File: ZLGitHubClient/ZLGitHubClient/Class/UI/ZLUI/profile/ZLWebContentController/viewModel/ZLWebContentViewModel.m
//
//  ZLWebContentViewModel.m
//  ZLGitHubClient
//
//  Created by 朱猛 on 2019/8/28.
//  Copyright © 2019 ZM. All rights reserved.
//

#import "ZLWebContentViewModel.h"
#import <WebKit/WebKit.h>

@interface ZLWebContentViewModel()<ZLWebContentViewDelegate>

// view
@property(weak, nonatomic) ZLWebContentView * webContentView;

@property(strong, nonatomic) NSURL * url;

@end

@implementation ZLWebContentViewModel

- (void) dealloc
{
    //[self clearCookiesForWkWebView];
}

- (void) bindModel:(id _Nullable) targetModel andView:(UIView *) targetView
{
    if(![targetView isKindOfClass:[ZLWebContentView class]])
    {
        ZLLog_Warning(@"targetView is not ZLWebContentView,so return");
        return;
    }
    
    UIButton *rightButton = [UIButton buttonWithType:UIButtonTypeCustom];
    [rightButton setImage:[UIImage imageNamed:@"run_more"] forState:UIControlStateNormal];
    [rightButton setFrame:CGRectMake(0, 0, 60, 60)];
    [rightButton addTarget:self action:@selector(onAdditionButtonClickWithButton:) forControlEvents:UIControlEventTouchUpInside];
    ZLBaseViewController *vc = (ZLBaseViewController *)self.viewController;
    vc.zlNavigationBar.rightButton = rightButton;
    
    self.webContentView = (ZLWebContentView *)targetView;
    self.webContentView.delegate = self;
    
    self.url = (NSURL *) targetModel;
        
    if(!self.url || !self.url.resourceSpecifier){
        ZLLog_Warning(@"targetModel is not a valid URL,so return");
        [ZLToastView showMessage:@"Invalid URL"];
        return;
    }
    
    if(!self.url.scheme || (![@"https" isEqualToString:self.url.scheme] && ![@"http" isEqualToString:self.url.scheme] )){
        if([[UIApplication sharedApplication] canOpenURL:self.url]){
            [[UIApplication sharedApplication] openURL:self.url options:@{} completionHandler:nil];
            return;
        }
    }
    
    if(!self.url.scheme ){
        NSURL *url = [[NSURL alloc] initWithString:[NSString stringWithFormat:@"http://%@",self.url.resourceSpecifier]];
        self.url = url;
    }
    
    NSMutableURLRequest * request =[NSMutableURLRequest requestWithURL:self.url cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:60];
    [self.webContentView.webView loadRequest:request];
}



- (void) onAdditionButtonClickWithButton:(UIButton *) button {
    
    NSURL *url = self.webContentView.webView.URL;
    
    UIActivityViewController *activityVC = [[UIActivityViewController alloc] initWithActivityItems:@[url] applicationActivities:nil];
    activityVC.popoverPresentationController.sourceView = button;
    activityVC.excludedActivityTypes = @[UIActivityTypeMessage,UIActivityTypeMail,UIActivityTypeOpenInIBooks,UIActivityTypeMarkupAsPDF];

    activityVC.completionWithItemsHandler = ^(UIActivityType  _Nullable activityType, BOOL completed, NSArray * _Nullable returnedItems, NSError * _Nullable activityError) {
        if (completed) {
        }else {
        }
    };
    [self.viewController presentViewController:activityVC animated:YES completion:nil];
    
}

#pragma mark - ZLWebContentViewDelegate

- (void)webView:(WKWebView * _Nonnull)webView navigationAction:(WKNavigationAction * _Nonnull)navigationAction decisionHandler:(void (^ _Nonnull)(WKNavigationActionPolicy))decisionHandler {
    decisionHandler(WKNavigationActionPolicyAllow);
}


- (void)webView:(WKWebView * _Nonnull)webView navigationResponse:(WKNavigationResponse * _Nonnull)navigationResponse decisionHandler:(void (^ _Nonnull)(WKNavigationResponsePolicy))decisionHandler { 
    decisionHandler(WKNavigationResponsePolicyAllow);
}


- (void)webView:(WKWebView *)webView didReceiveServerRedirectForProvisionalNavigation:(null_unspecified WKNavigation *)navigation{
    NSURL *url = webView.URL;
    if(!url.scheme || (![@"https" isEqualToString:url.scheme] && ![@"http" isEqualToString:url.scheme] )){
        if([[UIApplication sharedApplication] canOpenURL:url]){
            [[UIApplication sharedApplication] openURL:url options:@{} completionHandler:nil];
            return;
        }
    }
    
}

- (void) onTitleChangeWithTitle:(NSString *) title{
    if(title){
        self.viewController.title = title;
    }
}


#pragma mark - 清除cookie

-(void) clearCookiesForWkWebView
{
    NSArray * types = @[WKWebsiteDataTypeCookies,WKWebsiteDataTypeSessionStorage];
    NSSet * set = [NSSet setWithArray:types];
 
    NSDate * date = [NSDate dateWithTimeIntervalSince1970:0];
    [[WKWebsiteDataStore defaultDataStore] removeDataOfTypes:set modifiedSince:date completionHandler:^{}];
}


@end

File: ZLGitHubClient/ZLGitHubClient/Class/UI/profile/ZLWebContentController/viewModel/ZLWebContentViewModel.swift
//
//  ZLWebContentViewModel.swift
//  ZLGitHubClient
//
//  Created by 朱猛 on 2022/5/21.
//  Copyright © 2022 ZM. All rights reserved.
//

import UIKit
import ZLUIUtilities
import ZLBaseExtension
import WebKit
import ZMMVVM

class ZLWebContentViewModel: ZMBaseViewModel {
    
    private(set) var url: URL?
    
    init(url: URL?) {
        super.init()
        
        guard let url = url else {
            ZLToastView.showMessage("Invalid URL")
            return
        }
        
        self.url = url
        
        if url.scheme == nil {
            self.url = URL(string: "https://\(url.absoluteString)")
        }
    }

    var webContentView: ZLWebContentView? {
        zm_view as? ZLWebContentView
    }
}


extension ZLWebContentViewModel: ZLWebContentViewDelegate {
    
    @objc func webView(_ webView: WKWebView, navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
        
        guard let urlStr = navigationAction.request.url?.absoluteString else {
            decisionHandler(.allow)
            return
        }
        
        if ZLCommonURLManager.openURL(urlStr: urlStr, completionHandler: { [weak self] result in
            if result, urlStr == self?.url?.absoluteString {
                self?.zm_viewController?.navigationController?.popViewController(animated: false)
            }
        }) {
            decisionHandler(.cancel)
            return
        }
        decisionHandler(.allow)
    }

    @objc func webView(_ webView: WKWebView, navigationResponse: WKNavigationResponse, decisionHandler: @escaping (WKNavigationResponsePolicy) -> Void) {
        decisionHandler(.allow)
    }

    @objc func webView(_ webView: WKWebView, didReceiveServerRedirectForProvisionalNavigation navigation: WKNavigation!) {
        
        if let url = webView.url {
            
            if let scheme = url.scheme,
               scheme == "https" ||
                scheme == "http" {
                return
            }
            
            if UIApplication.shared.canOpenURL(url) {
                UIApplication.shared.open(url, options: [:], completionHandler: nil)
            }
        }
            
    }

    @objc func onTitleChange(title: String?) {
        if let title = title {
            zm_viewController?.title = title 
        }
    }
}
